### PR TITLE
ci: add pre-commit hook

### PR DESCRIPTION
**Pull request by bingryan**
## Description

add git `pre-commit` hook 

## Usage

`step1`:  install pre-commit for you local repo

```
just git-pre-commit
``` 

`step2`: auto format submitted `**.rs` files for your `git commit xxxx`

### CODE REVIEW

**Comment by bingryan** [comment]
After several PRs, i sometimes submit some unformatted code, then get ` CI lint error `.  i think add `pre-commit` hook is useful for this case.

File: bin/pre-commit [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+set -euo pipefail
+
+REPO_ROOT_DIR=$(git rev-parse --show-toplevel)
+
+files=$(git diff --cached --name-only --diff-filter=ACMR | grep -Ei "\.rs$" || true)
+
+if [[ -n $files ]]; then
+    cargo fmt --all
+    git add -- $(echo "$files" | paste -s -d " " -)
```
Comment (raphjaph) [inline bin/pre-commit:11]:
```suggestion
    cargo fmt --check
```

Comment (bingryan) [reply bin/pre-commit:11]: This only automatically formats the submitted rust file.

for only `cargo fmt --check`, if there are unformatted files, you will need to execute `cargo fmt --all` and then `git add xxx` again.

**Comment by bingryan** [comment]
only format the submitted Rust file now

**Comment by raphjaph** [comment]
It's not a good idea to change files in a pre-commit hook. So we only want to check things not write to any files

**Review by raphjaph — APPROVED** [review]
lgtm

Made it very simple and generic for now, we can add more complex hooks later.

**Pull request merged**